### ISSUE TITLE
Ingress and Egress from Actor Systems (Sync<->Async Bridging)

### ISSUE THREAD

**Issue by autodidaddict** [actix/actix#428]
In my current codebase, I have a bunch of full OS threads (`thread::spawn`) that are running select loops to receive messages on channels and reply on other channels. This pattern is pervasive throughout my codebase. Each of these "message handling thread loops" looks a _lot_ like an actor. So, I tried my hand at converting the codebase from heavy OS threads to actix actors. The problem I'm having is that an actor system takes some inbound stimulus in order to start the massive chain of polling loops on futures, etc.

I've got an actor that encapsulates a struct that has a callback function over FFI (it's wrapping invocations into a WebAssembly module). The actor receives a request to invoke, needs to call into the wasm (which can, in turn, make outbound calls which will be satisfied by further actor calls), and then respond with the value. In a call chain where everything is aware of futures, this works fine. But, since the actor receives a message that makes an FFI call, the code inside the FFI (wasm) is synchronous and _expects a synchronous result_.

**tl;dr** at some vary small and crucial points in the application, I need to synchronously await the completion of a future from inside an actor's `handle` method that is not marked async (therefore `await` won't work). I've tried all kinds of shenanigans from using a `map` on the future to send the result out on a channel (doesn't work since nothing's polling the future that came from calling `send`), to using other runtimes to `block_on` the result (which just hangs forever).

I've asked on Gitter and Twitter for patterns and workarounds to this problem to no avail.

How does one await the result of some asynchronous actor invocation from inside a synchronous function that is _not_ marked as an `actix::main` or a `tokio::main` or some other runtime's entry point?

**Comment by autodidaddict** [comment]
In Akka, for example, we can do this simply by invoking `ask` on the actor in question and the framework will wait for a response from that actor (or fail after a configurable timeout). In Erlang/Elixir/OTP I can send any process a message and get a reply and the underlying plumbing is hidden from developers.

**Comment by Neopallium** [comment]
@autodidaddict So your Sync Actor code needs to make some async calls?

I think it should be possible to call `futures::executor::block_on(async move { ... async code ... })` to run the `async` code from  the current thread running your Sync Actor.

Here is a simple example of `async` code sending a request to a `SyncActor` and that actor calling `async` code.
```rust
use actix::prelude::*;
use futures::executor::block_on;

#[derive(Message)]
#[rtype(result = "u64")]
struct HandleRequest(pub u64);

async fn async_call(num: u64) -> u64 {
  num * num
}

struct SyncActor;

impl Actor for SyncActor {
    // It's important to note that you use "SyncContext" here instead of "Context".
    type Context = SyncContext<Self>;
}

impl Handler<HandleRequest> for SyncActor {
    type Result = u64;

    fn handle(&mut self, msg: HandleRequest, _: &mut Self::Context) -> Self::Result {
      println!("SyncActor: handle called({})", msg.0);
      let res = block_on(async move {
        println!("SyncActor: in async code: call async_call({})", msg.0);
        let res = async_call(msg.0).await;
        println!("SyncActor: in async code: res = {:?}", res);
        res
      });
      println!("SyncActor: handle: res = {:?}", res);

      res
    }
}

#[actix_web::main]
async fn main() -> Result<(), String> {
    // Start the SyncArbiter with 2 threads, and receive the address of the Actor pool.
    let addr = SyncArbiter::start(2, || SyncActor);

    // send 5 messages
    for n in 5..10 {
        // As there are 2 threads, there are at least 2 messages always being processed
        // concurrently by the SyncActor.
        println!("in main async code: Send async request: {}", n);
        let res = addr.send(HandleRequest(n)).await;
        println!("in main async code: Request({}): Reply({:?})", n, res);
    }
    println!("Finished");
    Ok(())
}
```

**Comment by autodidaddict** [comment]
My async actor is encapsulating something that gets an ffi call. That
function needs to return a discrete value and can't return a future.

So I need the callback to call an actor and wait for the reply.

**Comment by autodidaddict** [comment]
Maybe I'm missing an essential concept here. If I make my ffi-encapsulating
actor a sync actor like your example things might work. But wouldn't that
then mean all my async actors would have to jump through hoops to call the
sync one? Apologies for my confusion, it's been difficult getting some of
these concepts through some of the spotty docs.

**Comment by Neopallium** [comment]
All Actors (Sync/Async) send messages to call each other.  The send is async.

Since your ffi code doesn't support async, it will need to be in a `SyncActor` like the example I showed.  In that example the `SyncActor`'s handle method would call into the ffi code and if that code needed to make a call to `async fn ...`, then it would need to use `block_on` like in the example.

**Comment by autodidaddict** [comment]
Thanks for the example and the help! Much appreciated! I think it's finally starting to make a little sense and I'll  give this a try.

**Comment by Neopallium** [comment]
Try making a simple example without all of the ffi code.  Just create a dummy `NativeModule` that fakes the interface of your ffi code.  Keep everything in rust to make it easier to debug the design (it would be easier for the compiler to help).  Wrap the dummy `NativeModule` in a `SyncActor` and add an `AsyncActor` or just some `async` functions.  Then link the Actors together.   AsyncActor calls SyncActor calls AnotherAsyncActor.

**Comment by autodidaddict** [comment]
I'll definitely do that. Thanks again for the help.

**Comment by fakeshadow** [comment]
You can use a future-aware oneshot channel. Your FFI callback take ownership of the sender part and your actor await on the receiver part. When the callback get result from your wasm use the sender to pass it to actor.
The benefit of this is you don't block anything while sync actor would take up at least one dedicated OS thread which defeat the purpose of you using actor pattern.

**Comment by autodidaddict** [comment]
I'm already using one OS thread per wasm module in my case because all WebAssembly modules are (at least in the current spec) inherently single-threaded, and that makes their execution a bottleneck in an async event-loop environment where I have dozens or hundreds of these. I think if I create a SyncArbiter with as many threads as I expect to have actor (wasm) modules, then this might do the trick. I'll experiment this weekend @Neopallium and it might be exactly what I was looking for.

**Comment by autodidaddict** [comment]
So I've set up enough scaffolding to try and invoke the scenario that I want to verify:
1. Create an actor that wraps the wasm module
2. Send invocation to the actor
2a. That actor then calls into the wasm engine for low-level invoke
3. Callback comes back out of the engine, triggering my closure
4. My closure sends to another actor and awaits a reply

Creating the actor (inside a test marked `actix_rt::main`)
```rust
 let addr = SyncArbiter::start(2, move || {
        let buf = {
            let mut f = File::open("../wascc-host/examples/.assets/kvcounter.wasm").unwrap();
            let mut v = Vec::new();
            f.read_to_end(&mut v).unwrap();
            v
        };
        let claims = wascap::wasm::extract_claims(&buf).unwrap().unwrap().claims;
        let mid = MiddlewareProcessor::new().start();
        let bus = UnitActor::new().start(); // a fake actor that printlns and returns a fixed payload
        ModuleHost::try_new(bus.recipient(), mid, hk.clone(), claims, &buf, None).unwrap()
    });
```
Here, `ModuleHost` is the actor that's wrapping the WebAssembly module.
My closure (created inside `ModuleHost`'s `try_new` function) is invoked, which means the callback came out of the webassembly module. Then, I attempt to await the result of the invocation on another actor:

```rust
println!("About to await bus reply");
    let res = block_on(async move {
        bus.send(inv).await
    });
    println!("Got bus reply!");
```
This sits forever and never returns. The `UnitActor`'s `handle` function is never invoked. This sits and waits just like all of my other attempts at getting this scenario to work.

**Comment by autodidaddict** [comment]
I've also tried bumping the thread count to something higher than the number of actors @Neopallium and that hasn't helped either. Either I'm doing something really foolish, or sync actors can't really wait for actor replies with a `block_on` call ?

**Comment by autodidaddict** [comment]
OK so I think I figured out which foolish thing I was doing. I was starting both `bus` (the testing actor, which is async) and `ModuleHost` inside the SyncArbiter start. When I moved the bus start outside (which I assume started bus on the default `AsyncArbiter`), everything worked and I was able to get a reply from my unit actor.

**Comment by Neopallium** [comment]
Might not want to start the `MiddlewareProcessor` in the `SyncActor` either.  Normally you would spawn/start all the actors from the main thread during startup and just clone the addr channel to send them messages.

Are you spawning a `SyncActor` for each wasm module (.wasm file)?  You could create a `WasmExecutor` as a `SyncActor` and spawn X instances (at least one per cpu-core).  Each executor could maintain a pool of loaded wasm modules.  Actix would decide which .wasm file was needed to handle a request and send the request + wasm name to the `WasmExecutor` using it's channel.  One of the spawned executors would pickup the request + wasm name, check it's pool to see if it was loaded and call the 'on_request' function inside the wasm module to handle the request.

Depending on the wasm crate you are using it might be possible to keep a global shared pool of compiled wasm modules (the code part is immutable and should be thread safe).  The `WasmExecutor`s would only need to keep a local (non-shared) pool of instantiated wasm module instances (the memory/state mutable part).

I have made a wasm interpreter written in Rust (100% safe code).  Haven't released it yet, it was mostly just a proof of concept to see if it was possible.  Also I had designed it so it would be possible to fully support async calls into wasm and back out to Rust (async host fn calls into wasm module which calls a host async fn).  The wasm compiled code doesn't even need to know that it is calling an async function, since it looks the same as a blocking call to the wasm code.

**Comment by autodidaddict** [comment]
This is all great advice.. I'd been trying to figure out what to do with the middleware in this scenario.

How it works now is that there is a single root `Host`, and this host is capable of holding instances of [Actor modules](https://wascc.dev/docs/concepts/actors/) and [Capability Providers](https://wascc.dev/docs/concepts/cap_prov/) (plugins for NFRs like databases, HTTP server, HTTP client, blob storage, etc). In today's version of the runtime, each wasm module gets its own OS thread, each capability provider gets its own OS thread, and each binding between a provider and a wasm module gets another thread. This allows for the host to hold hundreds of actor modules, 10s of capability providers, and handle dispatching accordingly without any slow handlers corroding the rest of the host runtime. The wasm execution engine is [waPC](https://github.com/wapc/wapc-rust), an engine-pluggable (today supports `wasmtime` and `wasm3`) host that supports a stateless RPC mechanism between module and host. Middleware is run before and after wasm execution, and so needs to be at the very least awaitable, if not entirely synchronous (so it might not make sense as an actor?).

I like the `WasmExecutor` idea, but I worry that having less executors than wasm modules means that a request for actor B could end up being slowed down while the executor awaits an answer from actor A on a totally unrelated request. Providers are shared by multiple actors, and these actors can communicate with each other and with providers over a self-forming cluster (we call it a `lattice`). In other words, an actor inside a host process could be bound to a provider in a different process (hence my use of the `bus` actor in the test). So, we can have layouts like this:

Host A :  Actor A, Actor B, Actor C, Provider 1
Host B: Actor D, Actor B (horizontally scaled instance), Actor E
Host C: Actor F, Actor G, Provider 2
Host D: Provider 3

these all communicate with each other either in process or across the distributed bus, and each host needs to be able to scale to hundreds or possibly 1000s of actors.

**Comment by autodidaddict** [comment]
@Neopallium what kind of Arbiter/Actix actor layout would you recommend for the scenario where each root `Host` acts as a dispatcher between 100s of concurrent wasm modules and the plugins that grant them cloud capabilities?

**Comment by Neopallium** [comment]
The `WasmExecutor` idea was to allow multiple requests to the same wasm module to run concurrently.  The executor would be like a pool of worker threads.  Since most wasm runtimes (wasmer, wasmtime, wasm3, etc..) don't currently support async or pause/resume, each active request being handled by a wasm module will block the thread running it.  You would spawn X instances of the executor and that would be the maximum number of concurrent requests that could be blocked/running wasm code at the same time.  If most of the wasm modules spend a lot of time waiting on requests to other actors/providers/etc.., then you could spawn much more executors then num cpu-cores, since most would be blocked either waiting for a request or blocked waiting for a reply from another actor/provider.

If each wasm module (wasm actor in your design?) only has one `SyncActor` handling it, then it would be restricted to handling one concurrent request at a time.

The best solution would be if the wasm runtime supported async or pause/resume itself, or if the .wasm files were compiled with pause/resume support [1].  Then you wouldn't need the `SyncActor`s, since everything could be async, as long as a .wasm files didn't do a lot of compute work.

1. [WASM pause/resume](https://www.youtube.com/watch?v=qQOP6jqZqf8)

**Comment by autodidaddict** [comment]
Given that I don't think either `wasmtime` or `wasm3` supports pause/resume, the `WasmExecutor` idea sounds ideal.  

If I used a `SyncArbiter`, like:

```rust
let addr = SyncArbiter::start(3, || { WasmExecutor::new(&file_buffer) }); // Actor pool of 3 serving the same wasm funcs
```
If I'm reading the docs right, this will pin an instance of `WasmExecutor` to each of the threads created?

Then that would essentially fix the number of instances of that executor I'm running, and I think I'm okay with needing to specify the number of instances of a wasm module I want at creation time, but I'm curious how an active, running system can add more instances of different executors? Can actors call `SyncArbiter::start` or does that always have to be called outside the system?

In other words, I've got a struct called `Host` and I also have the message bus actor `MessageBus`. The host has an API to add a new actor, and the message bus can receive control plane messages that tell it to start a new actor. Can both of those things call `SyncArbiter::start` ?

**Comment by autodidaddict** [comment]
Put another way, what I'm struggling with is what does my system have in it at process start time, and then how do I add wasm executors during steady-state? When the message bus (an async actor) receives a control plane message, can it just call `SyncArbiter::start( n || { WasmExecutor.... })` ? or will that interfere because the message bus already has its own arbiter (the default arbiter it got when I called `bus.start()`?)

**Comment by Neopallium** [comment]
I will try to outline the idea for `WasmExecutor` better.

At startup a pool of executors would be spawned (they don't know which wasm modules to load yet).
```rust
let num_workers = 10;  // should be at least num_cpus can be much higher to improve concurrent request handling.

let host_api = /* host api or other common state the wasm modules will need. */;

// start the pool of executors.  Each one will have a non-shared `WasmExecutor` instance.
// Each executor will have a non-shared pool of loaded wasm modules.
let executor_addr = SyncArbiter::start(num_workers, || { WasmExecutor::new(host_api.clone()) });

// later from other async http handlers or other actix actors send a `WasmHandleRequest` message to the executor handle.
// This request will be handled by one of the executor instances that is not busy (load balancing).
let reply = executor_addr.send(WasmHandleRequest {
  module_name: /* name of .wasm file that should handle the request. */
  /* other request parameters to be passed to the wasm module. */
}).await?:
println!("do something with the reply from the wasm module: relpy = {:?}", reply);

```

```rust
pub struct WasmHandleRequest {
  module_name: String, /* this can be the file name. */
  /* other request parameters to be passed to the wasm's handle_request function. */
}

pub struct WasmExecutorClearCache(bool);

pub struct WasmHandleRequestReply {
  /* reply from wasm `handle_request` function. */
}

/// WasmModule is whatever type the wasm runtime uses for loaded modules
pub struct WasmModule{}

pub struct WasmExecutor {
  loaded_modules: HashMap<String, WasmModule>,  // pool of loaded modules local to each instance of WasmExecutor.
  /* host_api and other state. */
}

impl WasmExecutor {
  pub fn new(host_api: /* .. */) -> Self {
    Self {
      loaded_modules: HashMap::new(),
      /* init other state. */
    }
  }

  fn get_module(&mut self, name: &str) -> &mut WasmModule {
    if !self.loaded_modules.contains(name) {
      let module = /* new wasm module, load & compile it. */;
      self.loaded_modules.insert(name, module);
    }
    // return mutable ref to loaded module.
    return self.loaded_modules.get_mut(name);
  }
}

impl Actor for WasmExecutor {
    type Context = SyncContext<Self>;
}

impl Handler<WasmHandleRequest> for WasmExecutor {
    type Result = Result<WasmHandleRequestReply, ()>;

    fn handle(&mut self, msg: WasmHandleRequest, _: &mut Self::Context) -> Self::Result {
      // get the module from cache or load it if this is the first request for it.
      let module = self.get_module(msg.module_name);
      // execute the wasm code to handle the request.  This code depends on the wasm runtime, so this is just fake place-holder code.
      let reply = module.call('handle_request', msg./* parameters for wasm */)?;
      // return the reply/response from the wasm request handler.
      Ok(reply)
    }
}

// You could have other messages to tell the executor to reload/clear it's cache of loaded modules or to notify it that a .wasm file should be reloaded.  This message can be sent by code that watches the .wasm files for changes to live-reload the wasm modules as needed.
// Might need a better way to send this.  I just realized that it wouldn't go to all instances of the executor.
impl Handler<WasmExecutorClearCache> for WasmExecutor {
    type Result = Result<bool, ()>;

    fn handle(&mut self, msg: WasmExecutorClearCache, _: &mut Self::Context) -> Self::Result {
      // clears the full cache.  It would be better to just reload only the changed wasm module.
      self.loaded_modules = HashMap::new();
      Ok(true)
    }
}

```

**Comment by Neopallium** [comment]
I haven't tried it, but it might be possible to dynamically spawn `SyncArbiter`s one for each wasm module.  I don't think it would be possible to dynamically change how many threads each wasm modules's SyncActor had spawned without recreating it.

Seems a bit heavy-weight to be spawning many `SyncArbiter`s.  With the above `WasmExecutor` design there would be a fixed pool of worker threads that would be tuned based on concurrent request needs and available OS resources.

The above executor design would also possibly load the same .wasm file multiple times (once for each worker thread).  If this costs too much (high memory usages because of large wasm modules), then it might be best to create a shared pool for loaded wasm files.  Look at the connection pooling crates (deadpool, r2d2), instead of holding a shared pool of live db connections, keep a pool of loaded wasm modules (maybe allow X copies of the same module, X>=1).  The pool would be shared with all instances of the `WasmExecutor`s instead of each having a non-shared `HashMap`.  When one executor gets a `WasmHandleRequest` message it would try to get the requested wasm module from the pool (this might block on a Mutex, if the maximum number of copies of that module are already being used).  Once the executor has finished handling the request the wasm module would be returned to the pool.

A shared pool like that would make it easier to live-reload the wasm modules when the .wasm file has changed.

**Comment by autodidaddict** [comment]
Ok, so I might've misunderstood. 

Your idea is to create `n` threads of the `WasmExecutor`, and hidden inside that executor is a shared reference to the pool of all loaded modules (instances of the `WapcHost`). We wouldn't necessarily use this pool to spin up multiple copies of the same wasm, just use it so that its `Handle<Invocation>` implementation picks the right target wasm based on the invocation target. I could make the number of Actor execution threads (wasmexecutor arbiter threads) an environment variable so it can be tweaked.

One weird Arbiter-related behavior I noticed is that my actor's `started` and `starting` functions don't get called when I start them via the SyncArbiter. Is that a bug? I see output from the factory function being called, and I see output when I call `send` on the arbiter address... but I don't see `starting`/`started` output.

**Comment by autodidaddict** [comment]
Looking at `deadpool`, it looks like I'd have to write my own because the pool doesn't have keyed items, and I'd need to retrieve modules by their keys (each one has a public key, which is part of the `target` field on invocations). Pooling works when you're storing and retrieving identical objects, but each of my host processes will hold hundreds of _unique_ webassembly modules (which is why the current version dedicates a full thread to each running module).

**Comment by autodidaddict** [comment]
And one last worry - neither of the underlying engines (`wasm3` or `wasmtime`) have `Sync` and `Send` on their structs, therefore the `WapcHost` struct (the one with the synchronous RPC call function on it) is also neither `Sync` nor `Send`.

**Comment by autodidaddict** [comment]
I went back and re-read your comment with the code (thank you so much for your patience with me, btw!). 
What confused me was it looks like your sample is using a "lazy" model - if the module hasn't been loaded at the time the invocation gets sent, then load it up. This is a great model but I think when it gets to a high number of wasm modules, having a non-shared cache will be a memory problem because with **200** actors in a host and **10** threads in the executor pool, after a long enough runtime, I'll have a total of **2000** modules loaded, where each of the non-shared caches has a copy of one of the **200** modules.

So, like you said, I think the optimal solution is a shared cache/pool of modules that are used by the set of executors for function calling. A call on a `WapcHost` is not mutable, so I can accomplish the shared pool with read locks on individual items in a shared `Arc<RwLock<HashMap<String, Mutex<WapcHost>>>`

**Comment by Neopallium** [comment]
I would first try it without the shared pool for `WapcHost`.  See if it is even needed.  Even if you have 1000 unique .wasm files and want to spawn 1000 worker threads for the executor, it might not require too much memory.

Just use a non-shared `HashMap` in the executor to store the loaded `WapcHost`.

**Comment by Neopallium** [comment]
I did a quick check of `wasmtime`.
A loaded wasm module has two parts:
1. `wasmtime::Module` immutable compiled wasm module.  This is thread-safe (Send and Sync).
2. `wasmtime::Instance` mutable instantiated wasm module. This is the non-thread-safe part (!Send, !Sync).  It is created from a `wasmtime::Module`, store and host api.  This is not thread-safe because it holds the memory that a spawned/started wasm module uses to run.

Only the first immutable part would be cacheable in a shared pool.  Most of the wasm runtimes I have seen have this same split between immutable compiled code part and mutable instance.  So you can load/compile a wasm module once and instantiate X copies all sharing one copy of the modules code.  This is like host binary files, you can start multiple copies of the same bin/exe file and the host OS will only need to keep one mmap copy of the binary file (read-only executable code pages).

**Comment by autodidaddict** [comment]
>  Even if you have 1000 unique .wasm files and want to spawn 1000 worker threads for the executor, it might not require too much memory.

In the unshared scenario, the amount of memory used is the number of threads * number of active actors * actor file size + some small overhead. So in the 1,000 thread scenario, every single actor (2MB) is going to add an additional **2GB** of memory usage after enough invocations happen to warm the cache in every thread.

I think I have a way to make the `WapcHost` thread safe by doing "thread hiding" or "spawn hiding" - Instead of storing the instance as a field (thereby tainting the struct), I can spawn an async func with a library like `smol` that just receives invocations from a channel and replies on another - the channels can be stored as fields on the struct and to the outside world it looks the same. This would let me pass around clones to the executors because the clones would just be structs with a couple channels on it, the original thread would be sitting in the instance cache somewhere.

**Comment by autodidaddict** [comment]
Given some of these options, is there any merit to just starting an actor that wraps a `WapcHost` in its own 1-thread arbiter? I'm just trying to figure out whether it's worth some of the extra hoops that I have to jump through in order to maintain this pool of executors and a cache of host instances all so I can have a different number of threads than I do hosts.

If I have 100 actors and 3 capabilities, that's only 103 threads, plus the event loop thread for all the remaining async actors. In this option, each actor can have a nicely contained, mutable wasm instance, no possibility of duplicating the memory storage for any wasm module. There would also be no need to introduce the complexity of pooling wasm/wapc hosts.

**Comment by Neopallium** [comment]
Yes.  That might be the best way forward.  Spawn a `SyncArbiter` for each `WapcHost`.  Store the arbiter address in a `HashMap` wasm name -> arbiter addr.  Should still be able to dynamically load new wasm modules as needed (starting new `SyncArbiter`s).

**Comment by fakeshadow** [comment]
With tokio 1.4 `Handle::block_on` has been added. So you can do something like this now start from actix 0.11

```rust
use actix::prelude::*;

#[actix::main]
async fn main() {
    let _addr = MyAct::start(MyAct(0));

    let handle = tokio::runtime::Handle::current();
    let (tx, rx) = tokio::sync::oneshot::channel();

    let handle = std::thread::spawn(move || {
        std::thread::sleep(std::time::Duration::from_secs(1));
        let result = handle.block_on(async {
            tokio::time::sleep(std::time::Duration::from_secs(3)).await;
            let _ = tx.send(());
            true
        });
        println!("block_on resolved");
        result
    });

    rx.await.unwrap();

    assert!(handle.join().unwrap());
}


struct MyAct(usize);

impl Actor for MyAct {
    type Context = Context<Self>;

    fn started(&mut self, ctx: &mut Self::Context) {
        ctx.run_interval(std::time::Duration::from_secs(1), |act, _| {
            act.0 += 1;
            println!("actor ticking, state: {}", act.0);
        });
    }
}
```

So I think this can be considered resolved issue and please ping or reply if you feel this issue should be still actively discussed.
